torch/_inductor/compile_fx.py

```diff
@@ -792,254 +798,355 @@ def codegen_and_compile(
-                            "stack_trace", None
-                        ):
-                            disable = f"{disable} Found from {stack_trace}\n"
-                        V.graph.disable_cudagraphs_reason = disable
+            return compiled_graph
 
-                if V.aot_compilation is True:
-                    return compiled_fn
 
-                if cudagraphs and not V.graph.disable_cudagraphs_reason:
-                    from torch._inductor.cudagraph_utils import (
-                        check_lowering_disable_cudagraph,
-                    )
+def fx_codegen_and_compile(
+    gm: GraphModule,
+    example_inputs: Sequence[InputType],
+    cudagraphs: Optional[BoxedBool] = None,
+    static_input_idxs: Optional[Sequence[int]] = None,
+    is_backward: bool = False,
+    graph_id: Optional[int] = None,
+    cpp_wrapper: bool = False,
+    aot_mode: bool = False,
+    is_inference: bool = False,
+    layout_opt: Optional[bool] = None,
+    extern_node_serializer: Optional[Callable[[List[ExternKernelNode]], Any]] = None,
+) -> Union[CompiledFxGraph, str]:
+    static_input_idxs = static_input_idxs or ()
 
-                    V.graph.disable_cudagraphs_reason = (
-                        check_lowering_disable_cudagraph(V.graph.device_node_mapping)
-                    )
+    scheme: FxCompile = _InProcessFxCompile()
 
-                compiled_graph = CompiledFxGraph(
-                    compiled_fn,
-                    graph,
-                    gm,
-                    output_strides,
-                    V.graph.disable_cudagraphs_reason,
-                    metrics_helper.get_deltas(),
-                    counters["inductor"] - inductor_counters,
-                )
+    if not _debug_force_async and isinstance(scheme, _FxCompileSync):
+        # As a debugging optimization if the scheme supports sync then run it
+        # sync.
+        return scheme.codegen_and_compile_sync(
+            gm,
+            example_inputs,
+            cudagraphs=cudagraphs,
+            static_input_idxs=static_input_idxs,
+            is_backward=is_backward,
+            graph_id=graph_id,
+            cpp_wrapper=cpp_wrapper,
+            aot_mode=aot_mode,
+            is_inference=is_inference,
+            layout_opt=layout_opt,
+            extern_node_serializer=extern_node_serializer,
+        )
+
+    future = scheme.codegen_and_compile(
+        gm,
+        example_inputs,
+        cudagraphs=cudagraphs,
+        static_input_idxs=static_input_idxs,
+        is_backward=is_backward,
+        graph_id=graph_id,
+        cpp_wrapper=cpp_wrapper,
+        aot_mode=aot_mode,
+        is_inference=is_inference,
+        layout_opt=layout_opt,
+        extern_node_serializer=extern_node_serializer,
+    )
 
-        return compiled_graph
+    try:
+        # We're already buried under an event loop - python won't let us nest
+        # event loops so trick it out.
```

Comment:
Is there a reference about this pattern? In FB there's some specific utilities for this situation but they're not readily available here

Reply:
I am a bit suspicious, because I read the discussion at https://www.internalfb.com/wiki/Python/Asyncio/Loop_Safety/#nested-eventloops-aka-aw and there's nothing about this

Reply:
This is a hack I found on a thread talking about the problems with nested event loops.  Unfortunately python async doesn't seem to have a "proper" way to have something that is async and callable from both sync and async functions (it seems silly to have to create a thread just to have the new thread have an event loop and have the main thread block on the new thread - but that seems to be the advice).  I'll pull the async version out for now and add it back in when I do the actual parallel eager/compile stuff and I have a better idea of what I want the final interface to look like.

Reply:
Yeahhhh when evaluating Python async hacks not only do you have to ask "does it work" but also "is it doing terrible things to the runtime characteristics of the existing loops" lol.

Reply:
In this particular case it's fine - because the async routine was internally sync - so it really was just driving a simple state machine. But I concede that a more complicated internal implementation (which we'd need for the RE implementation) might do more "interesting" (and potentially bad) stuff.